### PR TITLE
expand goto_programt and goto_functionst templates

### PR DESCRIPTION
**Pull request by kroening**
The goto_program_template and the goto_functions_template each only have one
instance, but make it much harder to read the code.

### CODE REVIEW

File: src/goto-programs/goto_program.h [hunk 1]
```diff
@@ -44,21 +81,601 @@ class goto_programt:public goto_program_templatet<codet, exprt>
...
+inline bool order_const_target(
+  const goto_programt::const_targett i1,
+  const goto_programt::const_targett i2)
+{
+  const goto_programt::instructiont &_i1=*i1;
+  const goto_programt::instructiont &_i2=*i2;
+  return &_i1<&_i2;
+}
+
+struct const_target_hash
```
Comment (tautschnig) [inline src/goto-programs/goto_program.h]: This needs `// NOLINTNEXTLINE(readability/identifiers)` in the line before.

File: src/goto-programs/goto_program.h [hunk 2]
```diff
@@ -44,21 +81,601 @@ class goto_programt:public goto_program_templatet<codet, exprt>
...
+    while(!l->is_end_function())
+      ++l;
+
+    return l->function;
+  }
+
+  static const irep_idt get_function_id(
+    const goto_programt &p)
+  {
+    assert(!p.empty());
```
Comment (tautschnig) [inline src/goto-programs/goto_program.h]: `PRECONDITION`

File: src/goto-programs/goto_program.h [hunk 3]
```diff
@@ -44,21 +81,601 @@ class goto_programt:public goto_program_templatet<codet, exprt>
...
+
+  template <typename Target>
+  std::list<Target> get_successors(Target target) const;
+
+  void compute_incoming_edges();
+
+  /// Insertion that preserves jumps to "target".
+  void insert_before_swap(targett target)
+  {
+    assert(target!=instructions.end());
```
Comment (tautschnig) [inline src/goto-programs/goto_program.h]: `PRECONDITION`

File: src/goto-programs/goto_program.h [hunk 4]
```diff
@@ -44,21 +81,601 @@ class goto_programt:public goto_program_templatet<codet, exprt>
...
+    target->swap(instruction);
+  }
+
+  /// Insertion that preserves jumps to "target".
+  /// The program p is destroyed.
+  void insert_before_swap(
+    targett target,
+    goto_programt &p)
+  {
+    assert(target!=instructions.end());
```
Comment (tautschnig) [inline src/goto-programs/goto_program.h]: `PRECONDITION`

File: src/goto-programs/goto_program.h [hunk 5]
```diff
@@ -44,21 +81,601 @@ class goto_programt:public goto_program_templatet<codet, exprt>
...
+
+  /// Clear the goto program
+  void clear()
+  {
+    instructions.clear();
+  }
+
+  targett get_end_function()
+  {
+    assert(!instructions.empty());
```
Comment (tautschnig) [inline src/goto-programs/goto_program.h]: `PRECONDITION`

File: src/goto-programs/goto_program.h [hunk 6]
```diff
@@ -44,21 +81,601 @@ class goto_programt:public goto_program_templatet<codet, exprt>
...
+    const auto end_function=std::prev(instructions.end());
+    assert(end_function->is_end_function());
```
Comment (tautschnig) [inline src/goto-programs/goto_program.h]: `CHECK_RETURN`

File: src/goto-programs/goto_functions.h [hunk 7]
```diff
@@ -15,12 +15,107 @@ Date: June 2003
 #define CPROVER_GOTO_PROGRAMS_GOTO_FUNCTIONS_H
 
 #include "goto_program.h"
-#include "goto_functions_template.h"
 
-class goto_functionst:public goto_functions_templatet<goto_programt>
+#include <ostream>
+#include <cassert>
```
Comment (tautschnig) [inline src/goto-programs/goto_functions.h:20]: Remove and use `util/invariant.h`

File: src/goto-programs/goto_program.cpp [hunk 8]
```diff
@@ -482,3 +483,148 @@ std::string as_string(
...
+
+  // number the targets properly
+  unsigned cnt=0;
+
+  for(auto &i : instructions)
+  {
+    if(i.is_target())
+    {
+      i.target_number=++cnt;
+      assert(i.target_number!=0);
```
Comment (tautschnig) [inline src/goto-programs/goto_program.cpp]: `CHECK_RETURN`

File: src/goto-programs/goto_program.cpp [hunk 9]
```diff
@@ -482,3 +483,148 @@ std::string as_string(
...
+  // check the targets!
+  // (this is a consistency check only)
+
+  for(const auto &i : instructions)
+  {
+    for(const auto &t : i.targets)
+    {
+      if(t!=instructions.end())
+      {
+        assert(t->target_number!=0);
```
Comment (tautschnig) [inline src/goto-programs/goto_program.cpp]: `DATA_INVARIANT`

File: src/goto-programs/goto_program.cpp [hunk 10]
```diff
@@ -482,3 +483,148 @@ std::string as_string(
...
+        assert(t->target_number!=instructiont::nil_target);
```
Comment (tautschnig) [inline src/goto-programs/goto_program.cpp]: `DATA_INVARIANT`

File: src/goto-programs/goto_program.h [hunk 11]
```diff
@@ -12,28 +12,65 @@ Author: Daniel Kroening, kroening@kroening.com
 #ifndef CPROVER_GOTO_PROGRAMS_GOTO_PROGRAM_H
 #define CPROVER_GOTO_PROGRAMS_GOTO_PROGRAM_H
 
+#include <cassert>
```
Comment (tautschnig) [inline src/goto-programs/goto_program.h:15]: Remove

File: src/goto-programs/goto_program.h [hunk 12]
```diff
@@ -44,21 +81,601 @@ class goto_programt:public goto_program_templatet<codet, exprt>
...
+    typedef std::list<const_targett> const_targetst;
+
+    /// The list of successor instructions
+    targetst targets;
+
+    /// Returns the first (and only) successor for the usual case of a single
+    /// target
+    targett get_target() const
+    {
+      assert(targets.size()==1);
```
Comment (tautschnig) [inline src/goto-programs/goto_program.h]: `DATA_INVARIANT`

**Review by tautschnig — CHANGES_REQUESTED** [review]
I very much support this move, just guidelines on how to make the linter happy.

**Comment by smowton** [comment]
Sounds good :) Rebase on latest develop may help with the Travis/pr timeout, as a recent change of mine reduced some needless crap in `__CPROVER_initialize`. The AppVeyor error on the other hand is real.

**Review by tautschnig — APPROVED** [review]

**Review by smowton — APPROVED** [review]

**Review by tautschnig — CHANGES_REQUESTED** [review]
Visual Studio/AppVeyor says that the `typename` in line 153 of lazy_goto_functions_map.h needs to be removed now that the class no longer is templated.

Edit: actually a fair bit of lazy_goto_functions_map.h could now be moved to a cpp file.

**Comment by thk123** [comment]
TG pointer bump: diffblue/test-gen#1503 required some tweaks but will post here once passes CI.

File: src/analyses/constant_propagator.cpp [hunk 13]
```diff
@@ -20,6 +20,8 @@ Author: Peter Schrammel
 #include <util/simplify_expr.h>
 #include <util/cprover_prefix.h>
 
+#include <langapi/language_util.h>
+
```
Comment (martin-cs) [inline src/analyses/constant_propagator.cpp:24]: What does this change have to do with the subject of the PR?

File: src/goto-instrument/wmm/goto2graph.cpp [hunk 14]
```diff
@@ -1301,7 +1301,7 @@ bool instrumentert::is_cfg_spurious(const event_grapht::critical_cyclet &cyc)
 
   /* now test whether this part of the code can exist */
   goto_functionst::function_mapt map;
-  goto_function_templatet<goto_programt> one_interleaving;
+  goto_functiont one_interleaving;
```
Comment (martin-cs) [inline src/goto-instrument/wmm/goto2graph.cpp:1304]: Yes; this shouldn't have ever been like this.

**Review by martin-cs — APPROVED** [review]
Yes; one instance templates are always a bit sus.  I /think/ that the use of typedefs keeps various out-of-tree patch sets safe but it'd be good to hear from maintainers of the better known ones.

File: src/analyses/constant_propagator.cpp (hunk 13 above)
Comment (kroening) [reply src/analyses/constant_propagator.cpp:24]: It was a missing header, which was hidden by the fact that goto_program_template.h included it.

**Review by tautschnig — APPROVED** [review]

**Pull request merged**